lib/delayed_job/threaded_worker.rb

```diff
@@ -0,0 +1,102 @@
+class ThreadedWorker < Delayed::Worker
+  def initialize(num_threads, options={}, grace_period_seconds=30)
+    super(options)
+    @num_threads = num_threads
+    @threads = []
+    @unexpected_error = false
+    @mutex = Mutex.new
+    @grace_period_seconds = grace_period_seconds
+  end
+
+  def start
+    # add quit trap as in QuitTrap monkey patch
+    trap('QUIT') do
+      Thread.new { say 'Exiting...' }
+      stop
+    end
+
+    trap('TERM') do
+      Thread.new { say 'Exiting...' }
+      stop
+      raise SignalException.new('TERM') if self.class.raise_signal_exceptions
+    end
+
+    trap('INT') do
+      Thread.new { say 'Exiting...' }
+      stop
+      raise SignalException.new('INT') if self.class.raise_signal_exceptions && self.class.raise_signal_exceptions != :term
+    end
+
+    say "Starting threaded delayed worker with #{@num_threads} threads"
+
+    @num_threads.times do |thread_index|
+      thread = Thread.new do
+        Thread.current[:thread_name] = "thread:#{thread_index + 1}"
+        threaded_start
+      rescue Exception => e # rubocop:disable Lint/RescueException
+        @mutex.synchronize do
+          say "Unexpected error: #{e.message}\n#{e.backtrace.join("\n")}", 'error'
```

Comment:
The mutex is only required for setting the `unexpected_error` attribute, or? Can the message be moved out?